### PR TITLE
[READY] make calculateDifficulty separate

### PR DESCRIPTION
**Pull request by winsvega**
NoProof support calculateDifficulty
Ethash support calculateDifficulty

### CODE REVIEW

**Comment by gumb0** [comment]
`BlockChainMainNetworkSuite/Mining_5_BlockFutureTime` fails everywhere

File: libethcore/SealEngine.h [hunk 1]
```diff
@@ -138,7 +138,13 @@ class NoProof: public eth::SealEngineBase
     static std::string name() { return "NoProof"; }
     static void init();
     void generateSeal(BlockHeader const& _bi) override;
+    void populateFromParent(BlockHeader& _bi, BlockHeader const& _parent) const override;
 };
 
+u256 calculateDifficulty(
```
Comment (gumb0) [inline libethcore/SealEngine.h]: Call it `calculateEthashDifficulty`

File: libethcore/SealEngine.h [hunk 2]
```diff
@@ -138,7 +138,13 @@ class NoProof: public eth::SealEngineBase
...
+    SealEngineFace const& sealEngine, BlockHeader const& _bi, BlockHeader const& _parent);
```
Comment (gumb0) [inline libethcore/SealEngine.h]: you don't need the whole `SealEngine`, pass `ChainParams` instead

File: libethcore/SealEngine.h (hunk 1 above)
Comment (winsvega) [reply libethcore/SealEngine.h]: Sounds weird calling EthashDifficulty from NoProof

File: libethcore/SealEngine.cpp [hunk 3]
```diff
@@ -151,3 +158,80 @@ u256 SealEngineBase::blockReward(u256 const& _blockNumber) const
     EVMSchedule const& schedule{evmSchedule(_blockNumber)};
     return chainParams().blockReward(schedule);
 }
+
+namespace dev
```
Comment (gumb0) [inline libethcore/SealEngine.cpp]: Better put everything in the file into these and remove `using namespace dev` `using namespace eth`

File: libethcore/SealEngine.h (hunk 1 above)
Comment (gumb0) [reply libethcore/SealEngine.h]: right, but that's what we need to do, I think it better expresses what's happening

Comment (winsvega) [reply libethcore/SealEngine.h]: is it also calculateEthashGasPrice ?

Comment (gumb0) [reply libethcore/SealEngine.h]: I'd leave `calculateGasLimit`, it doesn't seem to me to be inherent to Ethash
(could also go to the base class, but this global function is fine, too)

File: libethcore/SealEngine.h [hunk 4]
```diff
@@ -138,7 +138,13 @@ class NoProof: public eth::SealEngineBase
     static std::string name() { return "NoProof"; }
     static void init();
     void generateSeal(BlockHeader const& _bi) override;
+    void populateFromParent(BlockHeader& _bi, BlockHeader const& _parent) const override;
 };
 
+u256 calculateEthashDifficulty(
+    ChainOperationParams const& chainParams, BlockHeader const& _bi, BlockHeader const& _parent);
```
Comment (gumb0) [inline libethcore/SealEngine.h]: `_chainParams`

File: libethcore/SealEngine.h [hunk 5]
```diff
@@ -138,7 +138,13 @@ class NoProof: public eth::SealEngineBase
...
+
+u256 calculateGasLimit(ChainOperationParams const& chainParams, BlockHeader const& _bi,
```
Comment (gumb0) [inline libethcore/SealEngine.h]: `_chainParams`

File: test/unittests/libethashseal/EthashTest.cpp [hunk 6]
```diff
@@ -55,7 +55,8 @@ BOOST_AUTO_TEST_CASE(calculateDifficultyByzantiumWithoutUncles)
     header.setNumber(0x2001);
     header.setTimestamp(130);
 
-    BOOST_CHECK_EQUAL(ethash.calculateDifficulty(header, parentHeader), 999024);
```
Comment (gumb0) [inline test/unittests/libethashseal/EthashTest.cpp]: I'd leave this without a change, the tests here check `Ethash` class, we'd like to be sure that it calls the correct calculation

Comment (winsvega) [reply test/unittests/libethashseal/EthashTest.cpp]: how?
we've just removed calculateDifficulty from ethash into a separate function

Comment (winsvega) [reply test/unittests/libethashseal/EthashTest.cpp]: the test doesnt fail. that should be enough. it actually tests that function work as it was before

Comment (gumb0) [reply test/unittests/libethashseal/EthashTest.cpp]: Ah right, for a moment it seemed to me that `Ethash::calculateDifficulty` still existed. Ok then, this is fine

**Pull request merged**